File: cpp/DictParser.cpp
#include "DictParser.h"
#include <sstream>
#include <stdio.h>

enum DictParserParseStates {
    STATE_START = 0,
    STATE_CURLY,
    STATE_NAME_START,
    STATE_NAME_DATA,
    STATE_SIMPLE,
    STATE_BINARY_SIZE,
    STATE_BINARY_DATA,
    STATE_END
};

static void barf(std::string msg) {
    throw DictParser::parse_error(msg);
}

bool DictParser::getNextProperty(Property& property) {
    std::string name;
    std::string value;
    if (!getName(name)) return false; // EOF
    getValue(value);
    property.first = name;
    property.second = value;
    return true;
}

static int peekOrBarf(std::istream& stream, std::string msg) {
    int c = stream.peek();
    if (c == EOF) barf(msg);
    return c;
}

bool DictParser::getName(std::string& name) {
    if (state == STATE_END) return false;
    if (state == STATE_START || state == STATE_CURLY) matchOpenCurly();
    if (state == STATE_NAME_START) matchNameStart();
    return getNameValue(name);
}

void DictParser::matchNameStart() {
    if (state != STATE_NAME_START) { barf("invalid state: not at name start"); }

    int lookahead = peekOrBarf(input, "missing dict separator '}'");

    if (lookahead == '}') {
        input.get();
        state = STATE_END;
        return;
    }

    if (lookahead == '(' || lookahead == ':') barf("missing name");
    state = STATE_NAME_DATA;
}

void DictParser::matchOpenCurly() {
    int c = peekOrBarf(input, "unexpected EOF when reading open curly");
    if (c != '{') barf("missing initial '{'");
    input.get();
    state = STATE_NAME_START;
}

bool DictParser::getNameValue(std::string& name) {
    if (state == STATE_END) return false;
    if (state != STATE_NAME_DATA) barf("invalid state: not at name data");
    
    name.clear();
    while (input.good()) {
        char c;
        input.get(c);
        if (input.fail()) barf("unexpected EOF when reading property name");
        if (c == ':') {
            state = STATE_SIMPLE;
            break;
        } else if (c == '(') {
            state = STATE_BINARY_SIZE;
            break;
        }
        name.push_back(c);
    }
    return true;
}

void DictParser::getValue(std::string& value) {
    if (state == STATE_SIMPLE) { getSimpleValue(value); }
    if (state == STATE_BINARY_SIZE) { getBinaryValue(value); }
    matchNameStart();
}

void DictParser::getSimpleValue(std::string& value) {
    value.clear();
    while (input.good()) {
        char c;
        input.get(c);
        if (input.fail()) barf("unexpected EOF when reading simple property value");
        if (c == ';') { state = STATE_NAME_START; break; }
        value.push_back(c);
    }
}

void DictParser::getBinaryValue(std::string& value) {
    uint64_t size = getBinarySize();
    getBinaryData(value, size);
    matchBinaryPropertySeparator();
    state = STATE_NAME_START;
}

void DictParser::matchBinaryPropertySeparator() {
    int c = peekOrBarf(input, "unexpected EOF when reading binary property separator ';'");
    if (c != ';') barf("missing binary property separator ';'");
    input.get();
}

uint64_t DictParser::getBinarySize() {
    std::string digits;

    while (input.good()) {
        char c;
        input.get(c);
        if (input.fail()) barf("unexpected EOF when reading binary size");
        if (c == ')') break;
        if (c < '0' || c > '9') barf("size must be digits");
        digits.push_back(c);
    }
    if (digits.length() == 0) barf("size must be digits");
    matchBinaryNameSeparator();
    
    std::istringstream digitStream(digits);
    uint64_t size = 0;
    digitStream >> size;
    
    state = STATE_BINARY_DATA;
    return size;
}

void DictParser::getBinaryData(std::string& value, uint64_t size) {
    value.clear();
    for (uint64_t i=0; i<size; i++) {
        char c;
        input.get(c);
        if (input.fail()) barf("unexpected EOF when reading binary property value");
        value.push_back(c);
    }
}

void DictParser::matchBinaryNameSeparator() {
    int c = peekOrBarf(input, "unexpected EOF when reading binary value separator ':'");
    if (c != ':') barf("missing name separator ':'");
    input.get();
}

std::string DictParser::Property::name() const {
    return (*this).first;
}

std::string DictParser::Property::value() const {
    return (*this).second;
}

File: python/DictParser.py
import io
import sys

def val2str(i):
  if i == None: return "?"
  return str(i)

class Property(object):
    def __init__(self,name,value):
        self._name = name
        self._value = value
    
    def name(self):
        return self._name;
    
    def value(self):
        return self._value;

    def __repr__(self):
        return self.__str__()

    def __str__(self):
        return "<DictParser.Property: name='%s' value='%s'>" % (val2str(self._name), val2str(self._value))

    def __unicode__(self):
        return unicode(self.__str__())

class State:
    (START, CURLY, NAME_START, NAME_DATA, SIMPLE, BINARY_SIZE, BINARY_DATA, END) = range(8)

class DictParser(object):
    
    class ParseError(Exception):
        pass
    
    def __init__(self, inputStream):
        self.state = State.START
        self.input = io.BufferedReader(inputStream)
    
    def getNextProperty(self):
        name = self.parseName()
        if name == None: return None # EOF
        value = self.parseValue()
        return Property(name,value)
    
    #
    # Internal methods after this point
    #
    
    def parseName(self):
        if self.state == State.END: return None
        if self.state in [State.START, State.CURLY]: self.matchOpenCurly()
        if self.state == State.NAME_START: self.matchNameStart()
        return self.parseNameValue()
    
    def parseValue(self):
        val = None
        if   self.state == State.SIMPLE: val = self.parseSimpleValue()
        elif self.state == State.BINARY_SIZE: val = self.parseBinaryValue()
        self.matchNameStart()
        return val
    
    def parseNameValue(self):
        if self.state == State.END: return None
        if self.state != State.NAME_DATA: self.barf("invalid state: not at name data")
        
        nameBuf = []
        while True:
            c = self.readOrBarf("unexpected EOF when reading property name")
            if c == b':':
                self.state = State.SIMPLE
                break
            elif c == b'(':
                self.state = State.BINARY_SIZE
                break
            nameBuf.append(c)        
        return b''.join(nameBuf)
    
    def parseSimpleValue(self):
        valBuf = []
        while True:
            c = self.readOrBarf("unexpected EOF when reading simple property value")
            if c == b';':
                self.state = State.NAME_START
                break
            valBuf.append(c)
        return b''.join(valBuf)
    
    def parseBinaryValue(self):
        val = self.parseBinaryData( self.parseBinarySize() )
        self.matchBinaryPropertySeparator()
        self.state = State.NAME_START
        return val
    
    def parseBinarySize(self):
        digits = []
        digit_range = range(ord('0'),ord('9')+1)
        
        while True:
            c = self.readOrBarf("unexpected EOF when reading binary size")
            if c == b')': break
            if ord(c) not in digit_range:
                self.barf("size must be digits")
            digits.append(c)
        if len(digits) == 0: self.barf("size must be digits")

        self.matchBinaryNameSeparator()
        
        self.state = State.BINARY_DATA
        return int(b''.join(digits))
    
    def parseBinaryData(self, size):
        val = self.input.read(size)
        if len(val) != size: self.barf("unexpected EOF when reading binary property value")
        return val
    
    def matchOpenCurly(self):
        c = self.peekOrBarf("unexpected EOF when reading open curly")
        if c != b'{': self.barf("missing initial '{'")
        self.input.read(1)
        self.state = State.NAME_START
    
    def matchNameStart(self):
        if self.state != State.NAME_START: self.barf("invalid state: not at name start")
        
        lookahead = self.peekOrBarf("missing dict separator '}'")
        
        if lookahead == b'}':
            self.input.read(1)
            self.state = State.END
            return
        
        if lookahead in [b'(', b':']: self.barf("missing name")
        self.state = State.NAME_DATA
    
    def matchBinaryNameSeparator(self):
        c = self.peekOrBarf("unexpected EOF when reading binary value separator ':'")
        if c != b':': self.barf("missing name separator ':'")
        self.input.read(1)
    
    def matchBinaryPropertySeparator(self):
        c = self.peekOrBarf("unexpected EOF when reading binary property separator ';'")
        if c != b';': self.barf("missing binary property separator ';'")
        self.input.read(1)
    
    def peekOrBarf(self, message):
        buf = self.input.peek(1)
        if len(buf) < 1: self.barf(message)
        return buf[0]
    
    def readOrBarf(self, message):
        c = self.input.read(1)
        if len(c) != 1: self.barf(message)
        return c
    
    def barf(self,message):
        raise DictParser.ParseError(message)
